### ISSUE TITLE
S3 policies are not applied correctly

### ISSUE THREAD

**Issue by rishm069** [minio/minio#13675]
Specified S3 policies are not applied and user gets "403 Access Denied" error while admin user can get access without any issues. MinIO is deployed on VM and accessed from k8s cluster. 

## Expected Behavior

Buckets are accessible by user with corresponding policies.

## Current Behavior

"403 Access Denied" error is returned for any user's action.

## Possible Solution
N/A

## Steps to Reproduce (for bugs)

1. Set up user with the following configuration:
```
# User
resource "minio_iam_user" "user_archiver" {
  name = "archiver"
}

# Bucket
resource "minio_s3_bucket" "bucket_logs_archive" {
  bucket = "logs-archive"
  acl    = "public-read"
}

# Policies
data "minio_iam_policy_document" "policydoc_logs_archive_deployers" {
  statement {
    actions = [
      "s3:DeleteBucketPolicy",
      "s3:GetBucketPolicy",
      "s3:ListAllMyBuckets",
      "s3:ListBucket",
      "s3:PutBucketPolicy"
    ]

    resources = [
      "arn:aws:s3:::logs-archive",
    ]
  }

  statement {
    actions = [
      "s3:AbortMultipartUpload",
      "s3:DeleteObject",
      "s3:GetObject",
      "s3:ListMultipartUploadParts",
      "s3:PutObject"
    ]

    resources = [
      "arn:aws:s3:::logs-archive/*",
    ]
  }
}

resource "minio_iam_policy" "policy_logs_archive_deployers" {
  name   = "policy_logs_archive_deployers"
  policy = data.minio_iam_policy_document.policydoc_logs_archive_deployers.json
}

resource "minio_iam_user_policy_attachment" "policy-attachment_logs_archive_deployers" {
  user_name   = (minio_iam_user.user_archiver.id)
  policy_name = (minio_iam_policy.policy_logs_archive_deployers.id)
}
```
2. Try to access via `mc`

## Context

Here is the debug output from mc when I try to connect:
```
# ./mc --debug alias set minio-s3 http://minio.host archiver
Enter Secret Key: 
mc: <DEBUG> GET /probe-bucket-sign-fiv0c44t4qdl/?location= HTTP/1.1
Host: ******
User-Agent: MinIO (linux; amd64) minio-go/v7.0.15
Authorization: AWS4-HMAC-SHA256 Credential=archiver/20211117/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: *******
X-Amz-Date: 20211117T090739Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 403 Forbidden
Content-Length: 312
Accept-Ranges: bytes
Connection: keep-alive
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Wed, 17 Nov 2021 09:07:39 GMT
Server: nginx
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Origin
Vary: Accept-Encoding
X-Amz-Request-Id: ******
X-Content-Type-Options: nosniff
X-Xss-Protection: 1; mode=block

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied.</Message><BucketName>probe-bucket-sign-fiv0c44t4qdl</BucketName><Resource>/probe-bucket-sign-fiv0c44t4qdl/</Resource><RequestId>16B84A35BDB17954</RequestId><HostId>********</HostId></Error>mc: <DEBUG> Response Time:  9.981367ms

mc: <DEBUG> HEAD /probe-bucket-sign-fiv0c44t4qdl/ HTTP/1.1
Host: *****
User-Agent: MinIO (linux; amd64) minio-go/v7.0.15
Authorization: AWS4-HMAC-SHA256 Credential=archiver/20211117/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: ****
X-Amz-Date: 20211117T090739Z

mc: <DEBUG> HTTP/1.1 403 Forbidden
Accept-Ranges: bytes
Connection: keep-alive
Content-Security-Policy: block-all-mixed-content
Date: Wed, 17 Nov 2021 09:07:39 GMT
Server: nginx
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Origin
Vary: Accept-Encoding
X-Amz-Request-Id: *****
X-Content-Type-Options: nosniff
X-Xss-Protection: 1; mode=block
Content-Length: 0

mc: <DEBUG> Response Time:  8.389015ms

Added `minio-s3` successfully.
```

The intention is to use Minio as S3 storage for Grafana Loki (this application also gets 403 error, but works under admin user without any issues).

## Regression
N/A
## Your Environment

* Version used (`minio --version`): RELEASE.2021-08-05T22-01-19Z
* Operating System and version (`uname -a`): CentOS 7

**Comment by harshavardhana** [comment]
> # ./mc --debug alias set minio-s3 http://minio.host archiver
> Enter Secret Key: 
> mc: <DEBUG> GET /probe-bucket-sign-fiv0c44t4qdl/?location= HTTP/1.1
> Host: ******
> User-Agent: MinIO (linux; amd64) minio-go/v7.0.15
> Authorization: AWS4-HMAC-SHA256 Credential=archiver/20211117/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
> X-Amz-Content-Sha256: *******
> X-Amz-Date: 20211117T090739Z
> Accept-Encoding: gzip
> 
> mc: <DEBUG> HTTP/1.1 403 Forbidden
> Content-Length: 312
> Accept-Ranges: bytes
> Connection: keep-alive
> Content-Security-Policy: block-all-mixed-content
> Content-Type: application/xml
> Date: Wed, 17 Nov 2021 09:07:39 GMT
> Server: nginx
> Strict-Transport-Security: max-age=31536000; includeSubDomains
> Vary: Origin
> Vary: Accept-Encoding
> X-Amz-Request-Id: ******
> X-Content-Type-Options: nosniff
> X-Xss-Protection: 1; mode=block
> 
> <?xml version="1.0" encoding="UTF-8"?>
> <Error><Code>AccessDenied</Code><Message>Access Denied.</Message><BucketName>probe-bucket-sign-fiv0c44t4qdl</BucketName><Resource>/probe-bucket-sign-fiv0c44t4qdl/</Resource><RequestId>16B84A35BDB17954</RequestId><HostId>********</HostId></Error>mc: <DEBUG> Response Time:  9.981367ms
> 
> mc: <DEBUG> HEAD /probe-bucket-sign-fiv0c44t4qdl/ HTTP/1.1
> Host: *****
> User-Agent: MinIO (linux; amd64) minio-go/v7.0.15
> Authorization: AWS4-HMAC-SHA256 Credential=archiver/20211117/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
> X-Amz-Content-Sha256: ****
> X-Amz-Date: 20211117T090739Z
> 
> mc: <DEBUG> HTTP/1.1 403 Forbidden
> Accept-Ranges: bytes
> Connection: keep-alive
> Content-Security-Policy: block-all-mixed-content
> Date: Wed, 17 Nov 2021 09:07:39 GMT
> Server: nginx
> Strict-Transport-Security: max-age=31536000; includeSubDomains
> Vary: Origin
> Vary: Accept-Encoding
> X-Amz-Request-Id: *****
> X-Content-Type-Options: nosniff
> X-Xss-Protection: 1; mode=block
> Content-Length: 0
> 
> mc: <DEBUG> Response Time:  8.389015ms

There is no error here at the end - alias is set correctly what is the issue? @rishm069 .

We auto probe the signature support so you don't have to worry about these HTTP messages returning 403, they are supposed to return 403 since we expect it to and it is a valid case in this situation.

**Comment by harshavardhana** [comment]
I think your issue report doesn't show anything about policies not getting applied correctly. Closing the issue for now feel free to update the issue with the relevant failure.

**Comment by rishm069** [comment]
@harshavardhana 
Sorry, I should've added this to the issue in the first place: here is the error when I try to `ls` buckets with the aforementioned alias (even though it is set correctly)
```
# ./mc --debug ls minio-s3
mc: <DEBUG> GET / HTTP/1.1
Host: *****
User-Agent: MinIO (linux; amd64) minio-go/v7.0.15 mc/RELEASE.2021-10-07T04-19-58Z
Authorization: AWS4-HMAC-SHA256 Credential=archiver/20211117/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: *******
X-Amz-Date: 20211117T093958Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 403 Forbidden
Content-Length: 226
Accept-Ranges: bytes
Connection: keep-alive
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Wed, 17 Nov 2021 09:39:59 GMT
Server: nginx
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Origin
Vary: Accept-Encoding
X-Amz-Request-Id: 16B84BF963A9699F
X-Content-Type-Options: nosniff
X-Xss-Protection: 1; mode=block

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied.</Message><Resource>/</Resource><RequestId>16B84BF963A9699F</RequestId><HostId>*****</HostId></Error>mc: <DEBUG> Response Time:  105.028063ms

mc: <ERROR> Unable to list folder. Access Denied.
 (1) ls.go:239 cmd.doList(..) Tags: [http://******]
 (0) client-s3.go:2021 cmd.(*S3Client).listInRoutine(..)
 Release-Tag:RELEASE.2021-10-07T04-19-58Z | Commit:***** | Host:**** | OS:linux | Arch:amd64 | Lang:go1.17.1 | Mem:2.7 MB/21 MB | Heap:2.7 MB/12 MB
```

Or I'm getting something wrong and there should be another policy added for the user to be able to `ls` from `mc`?

**Comment by rishm069** [comment]
Also, I checked policies as follows:
```
./mc admin user info minio-s3 archiver 
AccessKey: archiver
Status: enabled
PolicyName: policy_logs_archive_deployers
MemberOf: 
```
```./mc admin policy info minio-s3 policy_logs_archive_deployers
{
 "Version": "2012-10-17",
 "Statement": [
  {
   "Effect": "Allow",
   "Action": [
    "s3:DeleteBucketPolicy",
    "s3:GetBucketPolicy",
    "s3:ListAllMyBuckets",
    "s3:ListBucket",
    "s3:PutBucketPolicy"
   ],
   "Resource": [
    "arn:aws:s3:::logs_archive"
   ]
  },
  {
   "Effect": "Allow",
   "Action": [
    "s3:ListMultipartUploadParts",
    "s3:PutObject",
    "s3:AbortMultipartUpload",
    "s3:DeleteObject",
    "s3:GetObject"
   ],
   "Resource": [
    "arn:aws:s3:::logs_archive/*"
   ]
  }
 ]
}
```

**Comment by rishm069** [comment]
@harshavardhana shouldn't the user be able to list bucket(s) with these policies?
```
    "s3:ListAllMyBuckets",
    "s3:ListBucket",
```

**Comment by rishm069** [comment]
@harshavardhana I cannot re-open the issue, tho

**Comment by vadmeste** [comment]
@rishm069 you have this
```
Resource": [
    "arn:aws:s3:::logs_archive"
]
```
   
That's why you are not able to list all buckets

**Comment by harshavardhana** [comment]
> @rishm069 you have this
> 
> ```
> Resource": [
>     "arn:aws:s3:::logs_archive"
> ]
> ```
> 
> That's why you are not able to list all buckets

First you can't have buckets with `_` in them and there is a wrong bucket in the policy.

**Comment by rishm069** [comment]
@harshavardhana  Sorry, the whole issue was caused by Terraform bug: https://github.com/aminueza/terraform-provider-minio/issues/105 
tl;dr Policies are indeed not applied, but not because MinIO. It has been fixed and [merged](https://github.com/aminueza/terraform-provider-minio/pull/111) but not in the version that I used. 

I recreated policies manually and everything works now. Thanks for help!
